src/module.c

```diff
@@ -1838,18 +1838,49 @@ unsigned long long RM_GetClientId(RedisModuleCtx *ctx) {
 }
 
 
-/* Return the Name of the current client calling the currently active module
- * command. If client authentication is NULL will return "default"
+/*
+ * Return name of RedisModuleUser. Get current client RedisModuleUser via the
+ * RM_GetClientUserById API. If user reference in RedisModuleUser is null,
+ * "default" is returned.
  */
-RedisModuleString *RM_GetClientName(RedisModuleCtx *ctx) {
-    if (ctx->client == NULL) return NULL;
-    sds name = sdsnew(ctx->client->user ? ctx->client->user->name: "default");
+RedisModuleString *RM_GetUserName(RedisModuleCtx *ctx, RedisModuleUser *user) {
+    if (user == NULL) return NULL;
+    sds name = sdsnew(user->user ? user->user->name: "default");
     robj *str = createObject(OBJ_STRING, name);
     autoMemoryAdd(ctx,REDISMODULE_AM_STRING, str);
     return str;
 }
 
 
+/* Return reference of RedisModuleUser with the specified ID (that was
+ * previously obtained via the RedisModule_GetClientId() API). RedisModuleUser
+ * contains fake user which copy current client user's name and flags. If the
+ * client is not exists, NULL is returned. If client->user is NULL, user name
+ * will set to "default".
+ *
+ * User created here are not listed by the ACL command. Free the user using the
+ * function RM_FreeModuleUser()
+ */
+RedisModuleUser *RM_GetClientUserById(uint64_t id) {
+    client *client = lookupClientByID(id);
+    if (client == NULL) return NULL;
+
+    RedisModuleUser *new_user = zmalloc(sizeof(RedisModuleUser));
+    new_user->user = ACLCreateUnlinkedUser();
```

Comment:
i don't think we want to create a new user. i think we want our RedisModuleUser struct to hold a reference to the real `client->user`, and maybe add a flag to the RedisModuleUser struct to indicate that RM_FreeModuleUser shouldn't release it.

@yossigo WDYT?

Do we wanna add an API that extracts the user name based on client ID directly?
Or one that returns a RedisModuleUser and let's the module get the user name from these? (can be re-used with other APIs that work with a RedisModuleUser)

Reply:
@oranagra I don't see a lot of value in obtaining a reference to the user and then having accessor functions on the user reference, we can just operate directly on the client. i.e. `RM_GetClientUserName()` that returns a string.

Reply:
@yossigo I was thinking that maybe some day we'll add more and more APIs to get info from a user (like it's ACL rules), and maybe some day these users may not necessarily be used by a connected client. (e.g. some code that iterates on all the registered ACL users).
Do you think it should just be: `RedisModuleString *RM_GetClientUserNameById(uint64_t id)`?
(similar to `RM_GetClientInfoById` and `RM_GetClientCertificate`, which work on a Connection, and **not** on a User)

Reply:
@oranagra I think we can take the "risk" of some API duplication to keep things simple and avoid reference counting users, requiring callers to release the received user handle, etc.

Reply:
ok.
@OhBonsai sorry for the back and forth, let's go for `RM_GetClientUserNameById` (similar to your original suggestion, except for the fact it takes an ID as input rather than use the current client).
this way it can work with the REDISMODULE_EVENT_CLIENT_CHANGE event.

Reply:
> sorry for the back and forth

@oranagra .It doesnt matter. I learn a lot from the discussing😂

I have pushed a new commit. Please take a look😄